OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1413 @@
+        senderMap.keySet.foreach { k ⇒
+            if (returningIntentsMap.contains(k)) {
+                val intents = returningIntentsMap(k)
+                implicitIntents --= intents
+                val sender = senderMap(k)
+                intents.foreach { i ⇒
+                    sender.foreach { s ⇒
+                        val intent = i.clone()
+                        intent.caller = s._1
+                        intent.componentTypes += s._2
+                        intent.pc = s._3
+                        implicitIntents += intent
+                    }
+                }
+            } else {
+                //overapproximate (no source found)
+                senderMap(k).foreach{s =>
+                    val intent = new implicitIntent(s._1)
+                    intent.componentTypes += s._2
+                    intent.pc = s._3
+                    intent.action = undef
+                    intent.categories += undef
+                    intent.iData = undef
+                    implicitIntents += intent
+                }
+            }
+        }
+
+        //Intent matching: testing implicitIntents against comptype, actions and categories of intentFilters
+        for (intent ← implicitIntents) {
+            intent.componentTypes.foreach { compType ⇒
+                if (filterMap.contains(compType)) {
+                    if (filterMap(compType).contains(intent.action)) {
+                        filterMap(compType)(intent.action).keySet.foreach { categories ⇒
+                            if ((intent.categories.isEmpty && categories.isEmpty) || intent.categories.forall(categories.contains)) {
+                                if (intent.iData == undef) {
+                                    overApproximateEdges(intent, filterMap(compType)(intent.action)(categories))
+                                } else {
+                                    testData(intent, filterMap(compType)(intent.action)(categories))
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        //adding explicit intent edges
+        explicitIntents.foreach { i ⇒
+            val caller = i.caller
+            i.calledMethod.foreach { m ⇒
+                edges += ((caller, i.pc) -> m)
+            }
+        }
+    }
+
+    def testData(
+                  intent: implicitIntent,
+                  filters: ListBuffer[intentFilter]
+    ): Unit = {
+
+        for (filter ← filters) {
+            val types = filter.dataTypes
+            val schemes = filter.dataSchemes
+            val data = intent.iData
+            if (types.isEmpty && schemes.isEmpty) {
+                if (intent.iType.isEmpty && data.isEmpty) {
+                    matchFound(intent, filter)
+                }
+            } else {
+                if (data.nonEmpty) {
```

Comment:
This ends up extremely deeply nested. Can you maybe refactor the code to avoid the deep nesting levels? E.g., it should be possible to directly use else if here instead of an else that is followed by a nested if.